src/engines/fx/lpc/lpc.cpp

```diff
@@ -21,69 +24,73 @@ namespace otto::engines {
 
   LPC::LPC() : EffectEngine("LPC", props, std::make_unique<LPCScreen>(this)) {}
 
-  void LPC::makeSha(gsl::span<float> buffer, float sha_period, int lag)
+  void LPC::makeSha(gsl::span<float> buffer, int sha_period)
   {
     util::fill(buffer, 0.f);
-    for (int i = lag; i < buffer.size(); i += sha_period) {
+    for (int i = this->lag; i < buffer.size(); i += sha_period) {
       buffer[i] = 1;
     }
-    lag = sha_period - buffer.size()%sha_period;
+    this->lag = sha_period - (int)(buffer.size()%sha_period);
   }
 
-  audio::ProcessData<2> LPC::process(audio::ProcessData<2> data)
+  audio::ProcessData<2> LPC::process(audio::ProcessData<1> data)
   {
-    // TODO: Add processing
-    #define ORDER 25
-    for(int i=0;i<ORDER;++i) acov[i] = data[i];
-    acov = acovb(acov);
+    auto in = data.audio;
+    int N = (int) in.size();
+    for(int i=0;i<N;++i) acov[i] = in[i];
+    acovb(acov);
 
     // Vocal signals usually have fundamentals in the [20Hz; 600Hz] band
     const float f_min = 20;
     const float f_max = 600;
 
     const float sampling_freq = 44100;
 
-    const maxT = ceil(sampling_freq/f_min);
-    const minT = floor(sampling_freq/f_max);
+    const int maxT = (int) ceil(sampling_freq/f_min);
+    const int minT = (int) floor(sampling_freq/f_max);
 
-    auto pitch = detect_pitch(gsl::span<float> data, unsigned minT, unsigned maxT);
+    auto pitch = detect_pitch(gsl::span(in.begin(), in.size()), minT, maxT);
     // Estimate coefficients
-
-    solve_yule_walker(acov.subspan(0, ORDER), sigma_and_coeffs, scratch_buffer);
+    #define ORDER 25
+    solve_yule_walker(gsl::span(acov, ORDER), this->sigmaAndCoeffs, this->scratchBuffer);
 
     // Create the source from:
     // 1. White noise
     // 2. Sha
     // 3. Harmonized Sha
 
-    if (pitch !=0){
-      makeSha(data, pitch, lag);
-      for(auto&& s : data) s = s*0.8 + white()*0.2;
+    auto exciter = Application::current().audio_manager->buffer_pool().allocate();
+
+    if (pitch != 0){
+      makeSha(gsl::span(exciter.begin(), exciter.size()), pitch);
+      for(auto&& s : exciter) s = s*0.8f + white()*0.2f;
     }else{
-      for(auto&& s : data) s = white()*0.5;
+      for(auto&& s : exciter) s = white()*0.5f;
     }
 
-    // copy in buffer
-    auto data_out = data;
+    auto out = Application::current().audio_manager->buffer_pool().allocate_multi<2>();
```

Comment:
It might be preferable to modify `exciter` in place, and then only allocate one extra audio buffer afterwards, and copying from `exciter` into `out2` using `util::copy(exciter, out2.begin())`. Then you should just be able to `return {exciter, out2}`

This will save us an audio buffer, and two loops over separate data are generally better than one